new stats to count resources

lohanidamodar opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to appwrite here: https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md

Happy contributing!

-->

## What does this PR do?

- [x] Refactor usage to use time-series data for all the metrics
- [x] Deprecate database and aggregator calculator for metrics
- [x] Deprecate maintenance by deleting the usage data (we need to save usage data longer)

@TorstenDittmann, this needs to migration to migrate all 30m stats stats into 1h stats by combining every two row.

## Test Plan

Existing usage test

## Related PRs and Issues

(If this PR is related to any other PR or resolves any issue or related to any issue, link all related PR and issues here.)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

YES

shimonewman commented on app/controllers/shared/api.php:
```diff
@@ -404,7 +478,6 @@
         if (
             App::getEnv('_APP_USAGE_STATS', 'enabled') == 'enabled'
             && $project->getId()
-            && $mode !== APP_MODE_ADMIN // TODO: add check to make sure user is admin
```
Removal of the !(admin) is intended?

lohanidamodar replied:
Yes indeed.

eldadfux commented on src/Appwrite/Usage/Calculators/TimeSeries.php:
```diff
@@ -14,7 +14,16 @@ class TimeSeries extends Calculator
     protected Database $database;
     protected $errorHandler;
     private array $latestTime = [];
```
Let's add a few docblocks in this class to better explain what we do here.

eldadfux commented on src/Appwrite/Usage/Stats.php:
```diff
@@ -80,7 +80,8 @@ public function getNamespace()
     {
         $projectId = $this->params['projectId'] ?? '';
-        $tags = ",projectId={$projectId},version=" . App::getEnv('_APP_VERSION', 'UNKNOWN');
+        $projectInternalId = $this->params['projectInternalId'];
```
Let's add a few docblocks in this class to better explain what we do here.

lohanidamodar commented on app/controllers/shared/api.php:
```diff
@@ -158,6 +160,78 @@
+        $dbForProject->on(Database::EVENT_DOCUMENT_CREATE, function ($event, Document $document) use ($usage) {
+            $collection = $document->getCollection();
+            switch ($collection) {
+                case 'users':
```
not sure I understand what you mean? We need to count different resources based on collection so I have this format.

lohanidamodar commented on app/controllers/shared/api.php:
```diff
@@ -158,6 +160,78 @@
+            }
+        });
+
+        $dbForProject->on(Database::EVENT_DOCUMENT_DELETE, function ($event, Document $document) use ($usage) {
```
based on whether it's delete or create, the value needs to be positive or negative, we might be able to extract the callback into a separate function.

eldadfux replied:
maybe we can pass an argument?

lohanidamodar replied:
refactored into single function

lohanidamodar commented on app/tasks/maintenance.php:
```diff
@@ -144,11 +143,11 @@ function notifyDeleteCache($interval)
         $abuseLogsRetention = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_ABUSE', '86400');
-        $usageStatsRetention30m = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_USAGE_30M', '129600'); //36 hours
-        $usageStatsRetention1d = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_USAGE_1D', '8640000'); // 100 days
+        $usageStatsRetention1h = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_USAGE_1H', '8640000'); //100 days
```
last we synced you said for now let's not document it, just use it. It's not even in environment variable

lohanidamodar commented:
@eldadfux 
- updated doc block comments
- renamed `_APP_MAINTENANCE_RETENTION_HOURLY_USAGE` to `_APP_MAINTENANCE_RETENTION_USAGE_HOURLY`
- renamed `$hourlyUsageStatsRetention` to `$usageStatsRetentionHourly`
- renamed `setHourlyUsageRetentionDatetime` to `setUsageRetentionHourlyDateTime`

lohanidamodar commented on src/Appwrite/Event/Delete.php:
```diff
@@ -56,26 +55,14 @@ public function setDatetime(string $datetime): self
      * @return self
      */
-    public function setDateTime1d(string $datetime): self
+    public function setDateTime1h(string $datetime): self
```
updated

lohanidamodar commented on app/workers/deletes.php:
```diff
@@ -105,7 +105,7 @@ public function run(): void
 
             case DELETE_TYPE_USAGE:
-                $this->deleteUsageStats($this->args['dateTime1d'], $this->args['dateTime30m']);
+                $this->deleteUsageStats($this->args['dateTime1d'], $this->args['dateTime1h']);
```
updated

lohanidamodar commented on app/tasks/maintenance.php:
```diff
@@ -144,11 +143,11 @@ function notifyDeleteCache($interval)
         $cacheRetention = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_CACHE', '2592000'); // 30 days
 
-        Console::loop(function () use ($interval, $executionLogsRetention, $abuseLogsRetention, $auditLogRetention, $usageStatsRetention30m, $usageStatsRetention1d, $cacheRetention) {
+        Console::loop(function () use ($interval, $executionLogsRetention, $abuseLogsRetention, $auditLogRetention, $usageStatsRetention1h, $cacheRetention) {
```
updated

lohanidamodar commented on app/config/variables.php:
```diff
@@ -188,7 +188,7 @@
             [
                 'name' => '_APP_USAGE_DATABASE_INTERVAL',
-                'description' => 'Interval value containing the number of seconds that the Appwrite usage process should wait before aggregating stats from data in Appwrite Database. The default value is 15 minutes.',
+                'description' => 'Deprecated since _ use _APP_USAGE_AGGREGATION_INTERVAL instead.',
```
set

lohanidamodar commented on app/config/variables.php:
```diff
@@ -179,7 +179,7 @@
             [
                 'name' => '_APP_USAGE_TIMESERIES_INTERVAL',
-                'description' => 'Interval value containing the number of seconds that the Appwrite usage process should wait before aggregating stats and syncing it to Appwrite Database from Timeseries Database. The default value is 30 seconds.',
+                'description' => 'Deprecated since _ use _APP_USAGE_AGGREGATION_INTERVAL instead.',
```
set

lohanidamodar commented on composer.json:
```diff
@@ -51,7 +51,7 @@
         "utopia-php/cli": "0.13.*",
         "utopia-php/config": "0.2.*",
-        "utopia-php/database": "0.26.*",
+        "utopia-php/database": "0.27.0 as 0.26.0",
```
Working on other depending libraries upgrading (abuse, audit)

lohanidamodar replied:
Updated with latest versions

lohanidamodar commented on tests/e2e/General/HTTPTest.php:
```diff
@@ -163,7 +163,8 @@ public function testSpecs()
             $response['body'] = json_decode($response['body'], true);
             $this->assertEquals(200, $response['headers']['status-code']);
-            $this->assertEmpty($response['body']['schemaValidationMessages']);
+            // looks like recent change in the validator
```
synced and updated

lohanidamodar commented on app/tasks/usage.php:
```diff
@@ -114,63 +110,28 @@ function getInfluxDB(Registry &$register): InfluxDatabase
-    ->param('type', 'timeseries', new WhiteList(['timeseries', 'database']))
     ->desc('Schedules syncing data from influxdb to Appwrite console db')
-    ->action(function (string $type) use ($register, $logError) {
+    ->action(function () use ($register, $logError) {
```
updated with DI

TorstenDittmann commented on app/config/collections.php:
```diff
@@ -2327,6 +2327,17 @@
                 'array' => false,
                 'filters' => [],
             ],
+            [
+                '$id' => ID::custom('buildTime'),
+                'type' => Database::VAR_INTEGER,
+                'format' => '',
+                'size' => 0,
+                'signed' => true,
+                'required' => false,
+                'default' => null,
+                'array' => false,
+                'filters' => [],
+            ],
```
Why did we add this? This should be part of the build itself 👍🏻

lohanidamodar replied:
fixed

TorstenDittmann commented on app/controllers/shared/api.php:
```diff
@@ -47,6 +48,47 @@
+            } elseif (strpos($collection, 'database_') === 0) {
+                $usage
+                    ->setParam('databaseId', $document->getAttribute('databaseId'));
+                if (strpos($collection, '_collection_') != false) {
```
```suggestion
                if (strpos($collection, '_collection_') !== false) {
```

TorstenDittmann commented on src/Appwrite/Usage/Stats.php:
```diff
@@ -102,12 +105,13 @@ public function submit(): void
             $value = $this->params[$metric] ?? 0;
-            if ($value >= 1) {
-                $this->statsd->increment($metric . $tags);
+            if ($value == 1 || $value == -1) {
```
```suggestion
            if ($value === 1 || $value === -1) {
```

TorstenDittmann commented on src/Appwrite/Usage/Stats.php:
```diff
@@ -124,13 +128,16 @@ public function submit(): void
         foreach ($dbMetrics as $metric) {
             $value = $this->params[$metric] ?? 0;
-            if ($value >= 1) {
+            if ($value == 1 || $value == -1) {
```
```suggestion
            if ($value === 1 || $value === -1) {
```

TorstenDittmann commented on src/Appwrite/Usage/Stats.php:
```diff
@@ -143,13 +150,16 @@ public function submit(): void
         foreach ($storageMertics as $metric) {
             $value = $this->params[$metric] ?? 0;
-            if ($value >= 1) {
+            if ($value >= 1 || $value == -1 || $value < -1) {
```
```suggestion
            if ($value !== 0) {
```
?

TorstenDittmann commented on src/Appwrite/Usage/Stats.php:
```diff
@@ -176,19 +186,30 @@ public function submit(): void
+            $this->statsd->count('project.{scope}.storage.size' . $tags, $deploymentSize + $storageSize);
+        }
+
+        if ($deploymentSize > 0 || $deploymentSize <= -1) {
```
```suggestion
        if ($deploymentSize !== 0) {
```

is the same right?
